---
title: "2.1_multiome_mm468"
author: "Anna Schwager"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    number_sections: yes
    toc: yes
    toc_float: yes
    code_folding: show
---

Load the global variables 
```{r source, message=F, warning=F, echo=F}
here::i_am("scripts/2.1_multiome_mm468.Rmd")
mainDir = here::here()
source(knitr::purl(file.path(mainDir,"scripts/global_variables.Rmd"), quiet=TRUE))
source(knitr::purl(file.path(mainDir, "scripts/functions.Rmd"), quiet=TRUE))
set.seed(123)
```

## Preparation and object creation
### Setting directories and input files
```{r set_inputs, message=F, warning=F}
inputDir_local = file.path(inputDir, "hg38", "MM468", "one_cell_multiome")

outputDir_local = file.path(outputDir, "2.1_multiome_mm468") ; if(!file.exists(outputDir_local)){dir.create(outputDir_local)}
outputDir_objects = file.path(outputDir_local, "objects") ; if(!file.exists(outputDir_objects)){dir.create(outputDir_objects)}
outputDir_plots = file.path(outputDir_local, "plots") ; if(!file.exists(outputDir_plots)){dir.create(outputDir_plots)}
outputDir_tables = file.path(outputDir_local, "tables") ; if(!file.exists(outputDir_tables)){dir.create(outputDir_tables)}

ld <- list.dirs(inputDir_local)
fragpaths <- list.files(ld[grepl("/.*fragmentFiles", ld)], full.names = TRUE)
fragpaths <- fragpaths[grepl(".*/h3k27me3/.*fragmentFiles/.*tsv.gz$", fragpaths)]
names(fragpaths) <- "mm468"

rnapaths <- ld[grepl("/10XlikeMatrix_umi", ld)]

rm(ld)

```

### Loading annotation
```{r annot, message=F, warning=F}
consensus_peaks_k27 <- toGRanges(file.path(annotDir, "MM468_peaks_h3k27me3.bed"), format="BED", header=FALSE)
```

###  Creating the object 
```{r create_object, message=F, warning=F, eval=F}
## loading RNA data
rna.data <- Read10X(data.dir = rnapaths)
rna_seurat <- CreateSeuratObject(counts = rna.data,
                             min.cells = 1,
                             min.features = -1,
                             project = "mm468")

## creating the multiome seurat objects 
seurat_list <- list()
for (i in names(fragpaths)){
    message(paste0("### Making fragment object for ", i))
    total_counts <- CountFragments(fragpaths[[i]])
    barcodes <- total_counts$CB
    frags <- CreateFragmentObject(path = fragpaths[[i]], cells = barcodes)
    
    message(paste0("### Making 50k bin matrix for ", i))
    bin50k_kmatrix = GenomeBinMatrix(
    frags,
    genome = hg38,
    cells = NULL,
    binsize = 50000,
    process_n = 10000,
    sep = c(":", "_"),
    verbose = TRUE)
    
    message(paste0("### Making peak matrix for ", i))
    peak_matrix = FeatureMatrix(
      frags,
      features = consensus_peaks_k27,
      cells = NULL,
      sep = c("-", "-"),
      verbose = TRUE)
    
    message(paste0("### Creating chromatin assay for ", i))
    chrom_assay <- CreateChromatinAssay(
      counts = bin50k_kmatrix,
      sep = c(":", "_"),
      genome = "hg38",
      fragments = fragpaths[[i]],
      min.cells = 1,
      min.features = -1)
    
    message(paste0("### Creating Seurat object for ", i))
    seurat <- CreateSeuratObject(
      counts = chrom_assay,
      assay = "bin_50k")
    
    message(paste0("### Adding peak assay for ", i))
    seurat[["peaks"]] <- CreateChromatinAssay(
    counts = peak_matrix, genome = "hg38")
    
    message(paste0("### Adding RNA assay for ", i))
    seurat <- AddMetaData(seurat, rna_seurat@meta.data)
    seurat[["RNA"]] <- rna_seurat@assays[["RNA"]]

    Annotation(seurat) <- annotations_hg38
    
    seurat <- AddMetaData(seurat, CountFragments(fragpaths[[i]]))
    seurat <- FRiP(object = seurat, assay = "peaks", total.fragments = "reads_count")
    seurat@meta.data[["orig.ident"]] <- i
    seurat_list[[i]] <- seurat
    
    rm(seurat)
    rm(frags)
    rm(total_counts)
    rm(barcodes)
    rm(fragments_per_cell)
    rm(bin50k_matrix)
    rm(peak_matrix)
    rm(chromatin_assay)
}

seurat <- seurat_list[[1]]

saveRDS(seurat, file.path(outputDir_objects, "seurat_mm468_step1.rds"))

```

```{r seurat_reload, message=F, warning=F}
seurat <- readRDS(file.path(outputDir_objects, "seurat_mm468_step1.rds"))
```


## QC
### Weighted histograms
```{r qc1, message=F, warning=F, fig.show="hold", out.width="50%"}

plot_weighted_hist(seurat) + ggtitle(print(seurat@meta.data[["orig.ident"]][1]))

```

### Filtering

```{r merge_filter, message=F, warning=F}
min_reads = 500

seurat$filtering <- ifelse(seurat$reads_count > min_reads, 'pass', 'fail')

```

### FrIP and N fragments plots

```{r qc2, message=F, warning=F, fig.show="hold", out.width="50%"}

p_frip <- VlnPlot(subset(seurat, subset = filtering == 'pass'), c("FRiP"),
                      group.by = "orig.ident", split.by = NULL, pt.size = 0) +
                      labs(title = "FRiP MM468 H3K27me3") +
                      stat_summary(fun.y = median, geom='point', size = 2, colour = "black") +
                      theme(legend.position = "none") +
                      scale_fill_manual(values = mypal[3])
print(p_frip)


p_count <- VlnPlot(subset(seurat, subset = filtering == 'pass'), c("reads_count"),
                       group.by = "orig.ident", split.by = NULL, pt.size = 0) +
                       labs(title = "Fragments per cell MM468 H3K27me3") +
                       stat_summary(fun.y = median, geom='point', size = 2, colour = "black") +
                       theme(legend.position = "none") +
                      scale_fill_manual(values = mypal[3])
print(p_count)

p_count_rna <- VlnPlot(subset(seurat, subset = filtering == 'pass'), c("nCount_RNA"),
                       group.by = "orig.ident", split.by = NULL, pt.size = 0) +
                       labs(title = "Unique RNA reads per cell MM468 H3K27me3") +
                       stat_summary(fun.y = median, geom='point', size = 2, colour = "black") +
                       theme(legend.position = "none") +
                      scale_fill_manual(values = mypal[1])
print(p_count_rna)

p_genes_rna <- VlnPlot(subset(seurat, subset = filtering == 'pass'), c("nFeature_RNA"),
                        group.by = "orig.ident", split.by = NULL, pt.size = 0) +
                        labs(title = "Unique genes per cell MM468 H3K27me3") +
                        stat_summary(fun.y = median, geom='point', size = 2, colour = "black") +
                        theme(legend.position = "none") +
                        scale_fill_manual(values = mypal[1])
print(p_genes_rna)

```

```{r save1, message=F, warning=F, eval=F}
ggsave(file.path(outputDir_plots, "frip.pdf"),
       plot = p_frip,
       device = "pdf",
       units = "mm",
       width = 100,
       height = 110)

ggsave(file.path(outputDir_plots, "n_fragments.pdf"),
       plot = p_count,
       device = "pdf",
       units = "mm",
       width = 100,
       height = 110)

ggsave(file.path(outputDir_plots, "n_rna_reads.pdf"),
       plot = p_count_rna,
       device = "pdf",
       units = "mm",
       width = 100,
       height = 110)

ggsave(file.path(outputDir_plots, "n_genes.pdf"),
       plot = p_genes_rna,
       device = "pdf",
       units = "mm",
       width = 100,
       height = 110)
```


### DNA to RNA reads scatter plots
```{r qc3, message=F, warning=F, fig.show="hold", out.width="50%"}
seurat <- subset(seurat, subset = filtering == "pass")

frags_to_rna_reads <- ggplot(seurat@meta.data, aes(x = reads_count, y = nCount_RNA, color = orig.ident)) +
                             geom_point(alpha = 1, stroke = 0, size = 2) +  
                             labs(x = "N unique DNA fragments",
                                  y =  "N unique RNA reads",
                                title = "N unique DNA fragments vs. N unique RNA reads") +
                              theme_classic() +
                              scale_color_manual(values = c(mypal[3], mypal[2]))

frags_to_genes <- ggplot(seurat@meta.data, aes(x = reads_count, y = nFeature_RNA, color = orig.ident)) +
                             geom_point(alpha = 1, stroke = 0, size = 2) +  
                             labs(x = "N unique DNA fragments",
                                  y =  "N unique genes",
                                title = "N unique DNA fragments vs. N unique genes") +
                              theme_classic() +
                              scale_color_manual(values = c(mypal[3], mypal[2]))

genes_to_rna_reads <- ggplot(seurat@meta.data, aes(x = nFeature_RNA, y = nCount_RNA, color = orig.ident)) +
                             geom_point(alpha = 1, stroke = 0, size = 2) +  
                             labs(x = "N unique genes",
                                  y =  "N unique RNA reads",
                                title = "N unique genes vs. N unique RNA reads") +
                              theme_classic() +
                              scale_color_manual(values = c(mypal[3], mypal[2]))
frags_to_rna_reads
frags_to_genes
genes_to_rna_reads

```

```{r save2, message=F, warning=F, eval=F}
ggsave(file.path(outputDir_plots, "frags_to_rna_reads.pdf"),
       plot = frags_to_rna_reads,
       device = "pdf",
       units = "mm",
       width = 100,
       height = 70)

ggsave(file.path(outputDir_plots, "genes_to_rna_reads.pdf"),
       plot = genes_to_rna_reads,
       device = "pdf",
       units = "mm",
       width = 100,
       height = 70)

ggsave(file.path(outputDir_plots, "grags_to_genes.pdf"),
       plot = frags_to_genes,
       device = "pdf",
       units = "mm",
       width = 100,
       height = 70)
```


